File: 001 Two Sum/TwoSum.py
class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[]}
    def twoSum(self, nums, target):
        dict = {}
        # dict: key is the actual num, value is a list of index
        for i in xrange(len(nums)):
            if nums[i] in dict:
                dict[nums[i]].append(i + 1)
            else:
                dict[nums[i]] = [i + 1]

        for num in nums:
            if target - num in dict:
                # for instance, target is 6, one num is 3
                if target - num == num:
                    if len(dict[num]) > 1:
                        return (dict[num][0], dict[num][1])
                else:
                    return (dict[num][0], dict[target - num][0])
        return None

File: 001 Two Sum/TwoSum.js
/**
* @param {number[]} nums
* @param {number} target
* @return {number[]}
*/
var twoSum = function(nums, target) {

    var obj = {};

    for (var i = 0; i < nums.length; i++){
        if(nums[i] in obj){
            obj[nums[i]].push(i + 1);
        }
        else {
            obj[nums[i]] = [i + 1];
        }
    }

    for (var i = 0; i < nums.length; i++){
        if (target - nums[i] in obj){
            if (target - nums[i] == nums[i]){
                if (obj[nums[i]].length > 1){
                    return [obj[nums[i]][0], obj[nums[i]][1]];
                }
            }
            else{
                return [obj[nums[i]][0], obj[target - nums[i]][0]];
            }
        }
    }
};
